enclave-runtime/src/rpc/author/mod.rs

```diff
@@ -195,102 +227,137 @@ fn map_top_error<P: TrustedOperationPool>(error: P::Error) -> RpcError {
 	.into()
 }
 
-impl<'a, P> AuthorApi<TxHash<P>, BlockHash<P>> for Author<'a, P>
+impl<TopPoolGetter>
+	AuthorApi<
+		TxHash<TopPoolGetter::TrustedOperationPool>,
+		BlockHash<TopPoolGetter::TrustedOperationPool>,
+	> for Author<TopPoolGetter>
 where
-	P: TrustedOperationPool + Sync + Send + 'static,
+	TopPoolGetter: GetTopPool + Sync + Send + 'static,
 {
 	fn submit_top(
 		&self,
 		ext: Vec<u8>,
 		shard: ShardIdentifier,
-	) -> FutureResult<TxHash<P>, RpcError> {
+	) -> PoolFuture<TxHash<TopPoolGetter::TrustedOperationPool>, RpcError> {
 		self.process_top(ext, shard, TopSubmissionMode::Submit)
 	}
 
 	/// Get hash of TrustedOperation
-	fn hash_of(&self, xt: &TrustedOperation) -> TxHash<P> {
-		self.pool.hash_of(xt)
+	fn hash_of(
+		&self,
+		xt: &TrustedOperation,
+	) -> Result<TxHash<TopPoolGetter::TrustedOperationPool>> {
+		self.execute_on_top_pool(|t| Ok(t.hash_of(xt)), Err)
 	}
 
 	fn pending_tops(&self, shard: ShardIdentifier) -> Result<Vec<Vec<u8>>> {
-		Ok(self.pool.ready(shard).map(|top| top.data().encode()).collect())
+		self.execute_on_top_pool(
+			|t| Ok(t.ready(shard).map(|top| top.data().encode()).collect()),
+			Err,
```

Comment:
Do we really need to be abstract over the error fn? Will this ever be something different?

Reply:
Sorry, I did not see it above. However, I believe we can do this more idiomatic, although I don't know exactly how currently.

Reply:
Because of the function using futures, yeah. I don't think it's ideal either, but it's the best I could come up with for now 😅 
However, as I'm now doing the next step of the refactoring, I think the whole top pool behind a mutex will change. Since I will just be able to inject a regular top pool into the author, and then the author is kept in a global container.